### DISCUSSION TITLE
Modern.js vs Remix

### DISCUSSION THREAD

**Discussion by AsuraKev** [web-infra-dev/modern.js#4872] [General]
Hi there,

I have been experimenting with Modern js and so far the experience has been good. The main selling point for me really is the micro frontend tooling support over remix especially in the upcoming module federation support which is what I am waiting on. 

apart from this why would you recommend modern js over remix? Asking this because I need to pick a framework that is very future promising. 

There are similarities between remix and modern js, given that modern js is inspired by it. 

Thanks 🙏

**Comment by yimingjfe** [comment]
Hi, I am a member of Modern.js team. I want to share some ideas from my perspective.

Firstly I want to say that Remix is a fantastic framework. Modern.js was heavily inspired by Remix.js in routing and SSR. At the same time, Modern.js' routing solution is based on react-router, which is created by the Remix team.

There are some thoughts when we creating Modern.js:

We uses webpack and Rspack as bundlers in Modern.js. We have seen the perfect ecosystem of webpack and we hope that such an ecosystem can continue to develop. We're trying to make this ecosystem faster than ever with the help of Rspack.

In the enterprise-level application scenarios we face, micro-frontends are a very important way for us to solve the scaling problem, so we have made first-class support for micro-frontends in Modern.js. This also includes support for Module Federation.abs

We see that CSR (client-side rendering), SSR (server-side rendering) and SSG（Static Site Generation）each have their suitable usage scenarios. In some cases, using CSR mode can be more efficient and have lower operating costs. Therefore, in Modern.js, We provide support for all these rendering modes, and allow developers to choose the render mode they needed, which makes the entire framework more progressive.

**Reply by AsuraKev** [reply]
Thanks for the reply, it’s very valuable. There are downsides of remix when I considered it such as:

- unable to deploy to Azure static website as SSG is not supported
- no Plans to support module federation which mean micro frontend is techically not possible
- SSR by default and no ways to make it an SPA
- Seems to use some custom bundler and plans to move to vite


I am planning some architecture work for our product and will most likely be in the direction of micro frontend. Hearing from the module federation author that rspack will support module federation soon is super exciting. 

There are few things to ask before betting our product on top of Modern JS. 

- will this framework be actively maintained for the upcoming few years? Given that react isn’t dead xD
- Has this framework been used in production enterprise level application? 
- as react router is developed by Remix, whenever remix releases a new feature of react router, will modern js review and adopt it?

thanks!

**Reply by yimingjfe** [reply]
Of course, we will actively maintain this project. Modern.js is not just an open-source project. In Bytedance, there are at least thousands of projects in production that use Modern.js.

In Modern.js, we lock the version of react-router, and we will regularly review and adopt the new version of react-router. Today, we created a [PR](https://github.com/web-infra-dev/modern.js/pull/4873) to upgrade react-router to 6.17.0.
